File: sdk/ctv/android-tv/src/main/kotlin/com/rivalapexmediation/ctv/network/AuctionClient.kt
package com.rivalapexmediation.ctv.network

import android.content.Context
import android.os.Build
import android.os.SystemClock
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import com.rivalapexmediation.ctv.ApexMediation
import com.rivalapexmediation.ctv.SDKConfig
import com.rivalapexmediation.ctv.consent.ConsentData
import com.rivalapexmediation.ctv.metrics.MetricsRecorder
import com.rivalapexmediation.ctv.util.Logger
import okhttp3.Call
import okhttp3.Callback
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException
import java.net.SocketTimeoutException
import java.util.*
import java.util.concurrent.TimeUnit

class AuctionClient(
    private val context: Context,
    private val config: SDKConfig,
    private val gson: Gson = Gson(),
) {
    private val client: OkHttpClient = OkHttpClient.Builder()
        .connectTimeout(config.requestTimeoutMs.toLong(), TimeUnit.MILLISECONDS)
        .readTimeout(config.requestTimeoutMs.toLong(), TimeUnit.MILLISECONDS)
        .callTimeout((config.requestTimeoutMs * 2L), TimeUnit.MILLISECONDS)
        .build()

    data class Result(
        val win: AuctionWin?,
        val error: LoadError? = null,
    )

    fun requestBid(
        placementId: String,
        adFormat: String,
        floorCpm: Double,
        consent: ConsentData?,
        callback: (Result) -> Unit,
    ) {
        val requestId = UUID.randomUUID().toString()
        val req = AuctionRequest(
            requestId = requestId,
            placementId = placementId,
            adFormat = adFormat,
            floorCpm = floorCpm,
            device = buildDevice(),
            user = buildUser(),
            app = buildApp(),
            consent = buildConsentMap(consent),
            signal = null,
        )
        val body = gson.toJson(mapOf("body" to req)).toRequestBody("application/json".toMediaType())
        val url = config.apiBaseUrl.trimEnd('/') + "/rtb/bid"
        val builder = Request.Builder().url(url).post(body)
            .header("Content-Type", "application/json")
        config.apiKey?.let { builder.header("Authorization", "Bearer $it") }
        val httpReq = builder.build()

        val start = SystemClock.elapsedRealtime()
        client.newCall(httpReq).enqueue(object: Callback {
            override fun onFailure(call: Call, e: IOException) {
                Logger.w("Auction HTTP error: ${e.message}", e)
                val err = if (e is SocketTimeoutException) LoadError.Timeout else LoadError.Network
                ApexMediation.recordSloSample(false)
                MetricsRecorder.recordRequest(SystemClock.elapsedRealtime() - start, err)
                callback(Result(null, err))
            }
            override fun onResponse(call: Call, response: Response) {
                response.use { res ->
                    try {
                        if (res.code == 204) {
                            ApexMediation.recordSloSample(true)
                            MetricsRecorder.recordRequest(SystemClock.elapsedRealtime() - start, LoadError.NoFill)
                            callback(Result(null, LoadError.NoFill)); return
                        }
                        if (!res.isSuccessful) {
                            // Map simple taxonomy
                            val err = mapStatusError(res.code)
                            ApexMediation.recordSloSample(false)
                            MetricsRecorder.recordRequest(SystemClock.elapsedRealtime() - start, err)
                            callback(Result(null, err)); return
                        }
                        val bodyStr = res.body?.string() ?: "{}"
                        val mapType = object: TypeToken<Map<String, Any>>(){}.type
                        val parsed: Map<String, Any> = try { gson.fromJson(bodyStr, mapType) } catch (_: Exception) { emptyMap() }
                        val winJson = when {
                            parsed.containsKey("response") -> gson.toJsonTree(parsed["response"]).toString()
                            else -> bodyStr
                        }
                        val win = gson.fromJson(winJson, AuctionWin::class.java)
                        val loadError = if (win?.creativeUrl.isNullOrBlank()) LoadError.NoFill else null
                        ApexMediation.recordSloSample(true)
                        MetricsRecorder.recordRequest(SystemClock.elapsedRealtime() - start, loadError)
                        if (loadError == null) {
                            callback(Result(win, null))
                        } else {
                            callback(Result(null, loadError))
                        }
                    } catch (e: Exception) {
                        Logger.w("Auction parse error: ${e.message}", e)
                        ApexMediation.recordSloSample(false)
                        val err = LoadError.Generic("parse_error")
                        MetricsRecorder.recordRequest(SystemClock.elapsedRealtime() - start, err)
                        callback(Result(null, err))
                    }
                }
            }
        })
    }

    private fun buildDevice(): Map<String, Any?> = mapOf(
        "platform" to "android",
        "osVersion" to Build.VERSION.RELEASE,
        "model" to Build.MODEL,
        "tv" to true,
    )

    private fun buildUser(): Map<String, Any?> = emptyMap()

    private fun buildApp(): Map<String, Any?> = mapOf(
        "id" to config.appId,
        "name" to context.applicationInfo.loadLabel(context.packageManager).toString(),
        "bundle" to context.packageName,
        "version" to try { context.packageManager.getPackageInfo(context.packageName, 0).versionName } catch (_: Exception) { null }
    )

    private fun mapStatusError(code: Int): LoadError {
        return when {
            code == 408 || code == 504 -> LoadError.Timeout
            code == 429 -> LoadError.Status("status_429", code)
            code == 401 -> LoadError.Status("status_401", code)
            code in 400..499 -> LoadError.Status("status_4xx", code)
            code in 500..599 -> LoadError.Status("status_5xx", code)
            else -> LoadError.Status("status_${code}", code)
        }
    }
}

File: sdk/ctv/tvos/Sources/CTVSDK/AuctionClient.swift
import Foundation

public struct AuctionWin: Codable, Equatable {
    public let requestId: String
    public let bidId: String
    public let adapter: String
    public let cpm: Double
    public let currency: String
    public let ttlSeconds: Int
    public let creativeUrl: String
    public let tracking: Tracking
    public struct Tracking: Codable, Equatable {
        public let impression: String
        public let click: String
        public let start: String?
        public let firstQuartile: String?
        public let midpoint: String?
        public let thirdQuartile: String?
        public let complete: String?
        public let pause: String?
        public let resume: String?
        public let mute: String?
        public let unmute: String?
        public let close: String?
    }
}

public final class AuctionClient {
    private let config: SDKConfig
    private let session: URLSession

    public init(config: SDKConfig, session: URLSession = URLSession(configuration: .default)) {
        self.config = config
        self.session = session
    }

    public struct Result {
        public let win: AuctionWin?
        public let error: LoadError?
        public var noFill: Bool { error == .noFill }
    }

    public func requestBid(placementId: String,
                           adFormat: String,
                           floorCpm: Double = 0,
                           consent: ConsentData?,
                           completion: @escaping (Result) -> Void) {
        let requestId = UUID().uuidString
        var body: [String: Any] = [
            "requestId": requestId,
            "placementId": placementId,
            "adFormat": adFormat,
            "floorCpm": floorCpm,
            "device": ["platform": "ios", "osVersion": ProcessInfo.processInfo.operatingSystemVersionString, "model": "AppleTV"],
            "app": ["id": config.appId]
        ]
        if let c = consent {
            var m: [String: Any] = [:]
            if let g = c.gdprApplies { m["gdpr"] = g ? 1 : 0 }
            if let t = c.tcfString, !t.isEmpty { m["gdpr_consent"] = t }
            if let u = c.usPrivacy, !u.isEmpty { m["us_privacy"] = u }
            if let cp = c.coppa { m["coppa"] = cp }
            body["consent"] = m
        }
        let envelope = ["body": body]
        guard let data = try? JSONSerialization.data(withJSONObject: envelope) else {
            completion(Result(win: nil, error: .generic("encode_error"))); return
        }
        let url = URL(string: config.apiBaseUrl.trimmingCharacters(in: CharacterSet(charactersIn: "/")) + "/rtb/bid")!
        var req = URLRequest(url: url, cachePolicy: .reloadIgnoringLocalCacheData, timeoutInterval: TimeInterval(config.requestTimeoutMs) / 1000.0)
        req.httpMethod = "POST"
        req.addValue("application/json", forHTTPHeaderField: "Content-Type")
        if let key = config.apiKey { req.addValue("Bearer \(key)", forHTTPHeaderField: "Authorization") }
        req.httpBody = data
        let startTime = Date()

        let task = session.dataTask(with: req) { d, r, e in
            func finish(_ win: AuctionWin?, _ error: LoadError?) {
                let success = (error == nil) || (error == .noFill)
                ApexMediation.shared.recordSloSample(success: success)
                let elapsed = Date().timeIntervalSince(startTime) * 1000.0
                MetricsRecorder.shared.recordRequest(durationMs: elapsed, error: error)
                completion(Result(win: win, error: error))
            }
            if let err = e as NSError? {
                let loadError: LoadError = err.domain == NSURLErrorDomain && err.code == NSURLErrorTimedOut ? .timeout : .network
                finish(nil, loadError)
                return
            }
            guard let http = r as? HTTPURLResponse else {
                finish(nil, .generic("no_response"))
                return
            }
            if http.statusCode == 204 {
                finish(nil, .noFill)
                return
            }
            guard http.statusCode >= 200 && http.statusCode < 300 else {
                finish(nil, self.mapStatusError(http.statusCode))
                return
            }
            guard let d = d else {
                finish(nil, .generic("empty"))
                return
            }
            do {
                if let obj = try JSONSerialization.jsonObject(with: d) as? [String: Any], let resp = obj["response"] {
                    let respData = try JSONSerialization.data(withJSONObject: resp)
                    let win = try JSONDecoder().decode(AuctionWin.self, from: respData)
                    finish(win, nil)
                    return
                }
                let win = try JSONDecoder().decode(AuctionWin.self, from: d)
                finish(win, nil)
            } catch {
                finish(nil, .generic("parse_error"))
            }
        }
        task.resume()
    }

    private func mapStatusError(_ code: Int) -> LoadError {
        if code == 408 || code == 504 { return .timeout }
        if code == 429 { return .status(code: code, label: "status_429") }
        if code == 401 { return .status(code: code, label: "status_401") }
        if (400..<500).contains(code) { return .status(code: code, label: "status_4xx") }
        if (500..<600).contains(code) { return .status(code: code, label: "status_5xx") }
        return .status(code: code, label: "status_\(code)")
    }
}
